File: Q0017_Letter_Combinations_of_a_Phone_Number/Solution.py
class Solution:
    res = []
    letter_map = ["",
                  " ",
                  "abc",
                  "def",
                  "ghi",
                  "jkl",
                  "mno",
                  "pqrs",
                  "tuv",
                  "wxyz"]

    # 这里要进行初始化操作
    def __init__(self):
        self.res = []

    def letterCombinations(self, digits: str):
        if digits == "":
            return self.res

        self.find_combination(digits, 0, "")
        return self.res

    def find_combination(self, digits: str, index: int, s: str):
        if len(digits) == index:
            self.res.append(s)
            return self.res

        c = digits[index]
        for i in self.letter_map[ord(c) - ord('0')]:
            self.find_combination(digits, index + 1, s + i)


if __name__ == '__main__':
    print(Solution().letterCombinations("234"))

    print(Solution().letterCombinations(""))

File: Q0017_Letter_Combinations_of_a_Phone_Number/Solution.java
package Q0017_Letter_Combinations_of_a_Phone_Number;

import java.util.ArrayList;
import java.util.List;

/**
 * @ Description:
 * @ Date: Created in 12:55 2019-04-18
 * @ Author: Anthony_Duan
 */
public class Solution {

    private List<String> res = new ArrayList<>();
    private final static String[] LETTER_MAP = {
            "",
            " ",
            "abc",
            "def",
            "ghi",
            "jkl",
            "mno",
            "pqrs",
            "tuv",
            "wxyz"
    };


    /**
     * 寻找 digits[index] 匹配的数字，获得digits[0...index]的翻译得到的一个字符串
     *
     * @param digits
     * @param index
     * @param s      保存了此时从digits[0...index-1]翻译到的一个字母字符串
     */
    private void findCombination(String digits, int index, String s) {

        if (index == digits.length()) {
            res.add(s);
            return;
        }
        char c = digits.charAt(index);

        assert (c >= '0' && c <= '9' && c != '1');
        String letters = LETTER_MAP[c - '0'];
        for (int i = 0; i < letters.length(); i++) {
            findCombination(digits, index + 1, s + letters.charAt(i));
        }

    }

    public List<String> letterCombinations(String digits) {

        if ("".equals(digits)) {
            return res;
        }

        findCombination(digits, 0, "");
        return res;
    }


    public static void main(String[] args) {
        String s = "";
        System.out.println(new Solution().letterCombinations(s));
    }
}
